File: references/weekend/phases.py
"""The multi-phase weekend pipeline: extract -> draft -> refine -> structure.

Extracted from weekend/llm.py to stay under the repo's 500-line cap.
weekend/llm.py re-exports these names, so existing imports keep working (the
same shim pattern as lib/config.py).

Class C2c: these four phases form a CARRIER CHAIN. Each one must pass
DATES/PRICE/AGES/LOCATION through verbatim -- the predecessor narrowed the
payload at every step ("name, location, description", then "name +
description"), so an event's real dates were gone two phases before the schema
that asked for them, and every date column rendered blank.
"""

import json


def _llm():
    """Lazy accessor for weekend.llm.

    weekend.llm re-exports these phase functions, so a module-level import here
    would be circular. The file already used function-local imports for
    weekend.prompts; this keeps that style.
    """
    from weekend import llm

    return llm

def _load_extract_signals():
    try:
        if _llm().EXTRACT_SIGNALS_PATH.exists():
            return json.loads(_llm().EXTRACT_SIGNALS_PATH.read_text())
    except Exception:
        pass
    return {}


def _save_extract_signals(signals):
    _llm().EXTRACT_SIGNALS_PATH.parent.mkdir(parents=True, exist_ok=True)
    _llm().EXTRACT_SIGNALS_PATH.write_text(json.dumps(signals, indent=2, sort_keys=True))


def _extract_group_prompt(lines, source_type):
    from weekend.prompts import build_source_extract_prompt

    return build_source_extract_prompt("\n".join(lines), source_type)


def extract_sources(raw_text, source_type, model_name="default", use_foundation=False):
    signals = _load_extract_signals()
    per_type = signals.setdefault(model_name or "default", {}).setdefault(source_type, {})
    batch_size = per_type.get("batch_size", _llm().DEFAULT_BATCH_SIZE)
    phase_key = f"extract_sources/{source_type}"

    lines = [line for line in raw_text.split("\n") if line.startswith("- ")]
    if not lines:
        return raw_text

    results = []
    streak = 0
    i = 0
    while i < len(lines):
        chunk = lines[i : i + batch_size]
        prompt = _extract_group_prompt(chunk, source_type)
        result = _llm()._call_llm(
            "",
            prompt,
            timeout=_llm().PHASE_TIMEOUT_EXTRACT,
            phase_key=phase_key,
            use_foundation=use_foundation,
        )

        if result:
            results.append(result)
            streak += 1
            i += batch_size
            if streak >= _llm().BATCH_GROWTH_STREAK_LIMIT and batch_size < _llm().MAX_BATCH_SIZE:
                batch_size += 1
                per_type["batch_size"] = batch_size
                _save_extract_signals(signals)
        else:
            per_type["timeouts"] = per_type.get("timeouts", 0) + 1
            batch_size = max(batch_size // 2, 1)
            per_type["batch_size"] = batch_size
            _save_extract_signals(signals)
            streak = 0
            if batch_size == 1:
                results.append(chunk[0])
                i += 1
                batch_size = per_type.get("batch_size", _llm().DEFAULT_BATCH_SIZE)

    if not results:
        return raw_text
    return "\n".join(results)


def draft_activities(
    weather_condensed,
    cleaned_sources,
    source_type,
    location,
    age_range,
    date_range,
    use_foundation=False,
    year=None,
):
    from weekend.prompts import build_draft_prompt

    prompt = build_draft_prompt(
        weather_condensed, cleaned_sources, source_type, location, age_range, date_range,
        year=year,
    )
    return _llm()._call_llm(
        "",
        prompt,
        timeout=_llm().PHASE_TIMEOUT_DRAFT,
        phase_key=f"draft_activities/{source_type}",
        use_foundation=use_foundation,
    )


def refine_draft(draft_text, use_foundation=False):
    from weekend.prompts import build_refine_prompt

    result = _llm()._call_llm(
        "",
        build_refine_prompt(draft_text),
        timeout=_llm().PHASE_TIMEOUT_REFINE,
        phase_key="refine_draft",
        use_foundation=use_foundation,
    )
    return result if result else draft_text


def structure_to_json(
    text, source_type, age_range, weather_condensed="", use_foundation=False, year=None
):
    from weekend.prompts import build_structure_system_prompt, build_structure_user_prompt

    sys_prompt = build_structure_system_prompt(
        source_type, age_range, weather_condensed, year=year
    )
    usr_prompt = build_structure_user_prompt(text)
    return _llm()._call_llm(
        sys_prompt,
        usr_prompt,
        timeout=_llm().PHASE_TIMEOUT_STRUCTURE,
        parse_json=True,
        phase_key=f"structure_to_json/{source_type}",
        use_foundation=use_foundation,
    )

File: rust/src/ztools/weekend/phases.rs
//! The multi-phase weekend pipeline: extract -> draft -> refine -> structure.
//!
//! Ported from `weekend/phases.py`. Class C2c: these phases form a CARRIER
//! CHAIN. Each one must pass DATES/PRICE/AGES/LOCATION through verbatim -- the
//! predecessor narrowed the payload at every step, so an event's real dates
//! were gone two phases before the schema that asked for them, and every date
//! column rendered blank.

use std::cmp;

use super::prompts::{
    render, CARRY_FIELDS, PHASE_DRAFT_TRANSIENT, PHASE_EXTRACT_EVENTS, PHASE_REFINE,
    PHASE_STRUCTURE_TRANSIENT_SYSTEM, PHASE_STRUCTURE_USER, PHASE_WEATHER_CONDENSE,
};

pub const WEATHER_PREVIEW_LIMIT: usize = 200;
/// Consecutive extract-batch failures that trip the pass-through breaker.
///
/// Without a breaker a dead model cost FOUR timeouts per
/// line (8 -> 4 -> 2 -> 1 -> raw) across every line of the corpus — at 300s
/// each, a 66-line corpus was a 22-hour run that the scheduler killed at 30
/// minutes, every week, with nothing to show.
pub const EXTRACT_FAILURE_BREAKER: usize = 3;
pub const DEFAULT_BATCH_SIZE: usize = 8;
pub const MAX_BATCH_SIZE: usize = 12;
pub const BATCH_GROWTH_STREAK_LIMIT: usize = 3;

/// The plan-level context every phase shares. Bundled so the phase signatures
/// stay narrow and one consumer cannot pass a plan year that disagrees with
/// another's date range.
#[derive(Clone)]
pub struct PlanContext {
    pub location: String,
    pub ages: String,
    pub date_range: String,
    pub year: i32,
    pub exclusions: String,
}

/// Resolve the weekend model against the active Osaurus roster if needed.
#[must_use]
pub fn resolve_weekend_model(base_url: &str, preferred_model: &str) -> String {
    let url = format!("{}/v1/models", base_url.trim_end_matches('/'));
    let Ok(client) = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(2))
        .build()
    else {
        return preferred_model.to_string();
    };

    let resp = match client.get(&url).send() {
        Ok(r) if r.status().is_success() => r.json::<serde_json::Value>().ok(),
        _ => None,
    };

    let Some(json) = resp else {
        return preferred_model.to_string();
    };

    let models: Vec<String> = json
        .get("data")
        .and_then(|d| d.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|m| m.get("id").and_then(|id| id.as_str()).map(String::from))
                .collect()
        })
        .unwrap_or_default();

    if models.is_empty() || models.iter().any(|m| m == preferred_model) {
        return preferred_model.to_string();
    }

    let pref_lower = preferred_model.to_lowercase();
    let family = if pref_lower.contains("qwen") {
        "qwen"
    } else if pref_lower.contains("gemma") {
        "gemma"
    } else if pref_lower.contains("muse") {
        "muse"
    } else if pref_lower.contains("raptor") {
        "raptor"
    } else {
        ""
    };

    if !family.is_empty() {
        if let Some(matched) = models.iter().find(|m| m.to_lowercase().contains(family)) {
            return matched.clone();
        }
    }

    models
        .first()
        .cloned()
        .unwrap_or_else(|| preferred_model.to_string())
}

/// Wake the weekend model with one tiny request, waiting the LOADING budget.
///
/// Every other call in this pipeline carries a generation timeout (120s /
/// 300s). A cold 25GB model can take minutes to page in, and a client that
/// disconnects mid-load makes the server cancel the load — so the next call
/// starts it again, and the model never becomes ready. One call with the
/// warm-up budget, before any phase, turns that livelock into one wait.
#[must_use]
pub fn warm_model(config: &crate::config::ZtoolsConfig) -> super::ModelHealth {
    let model = resolve_weekend_model(&config.osaurus_url, &config.weekend_model);
    let started = std::time::Instant::now();
    // The stall guard IS the budget here: no token until the model has
    // loaded is the expected shape of a cold start, not a stalled server.
    let outcome = crate::ztools::llm::chat(
        &crate::ztools::llm::ChatRequest {
            base_url: &config.osaurus_url,
            model: &model,
            system: None,
            user: "Reply with the single word: ready",
            json: false,
        },
        &crate::ztools::llm::ChatBudget {
            stall_secs: config.llm_warmup_timeout_secs,
            cap_secs: config.llm_warmup_timeout_secs,
            max_tokens: 8,
        },
    );
    let secs = started.elapsed().as_secs();
    match outcome {
        Ok(_) => super::ModelHealth::Ready { model, secs },
        Err(e) => super::ModelHealth::Unavailable {
            model,
            reason: format!(
                "no answer within {secs}s of a {}s warm-up budget: {}",
                config.llm_warmup_timeout_secs,
                first_line(&e.to_string())
            ),
        },
    }
}

fn first_line(text: &str) -> &str {
    text.lines().next().unwrap_or(text).trim()
}

/// One plain-text LLM call against the configured osaurus endpoint.
#[must_use]
pub fn call_llm_text(prompt: &str, config: &crate::config::ZtoolsConfig) -> Option<String> {
    let model = resolve_weekend_model(&config.osaurus_url, &config.weekend_model);
    crate::ztools::twitter::call_osaurus(
        &config.osaurus_url,
        &model,
        prompt,
        config.llm_extended_timeout_secs,
        config,
    )
    .ok()
    .filter(|s| !s.trim().is_empty())
}

/// One JSON LLM call returning the raw response value.
///
/// The answer is re-wrapped in the completion shape the parsers expect
/// (`choices[0].message.content`), so a streamed answer and a stubbed one
/// look the same to `parse_llm_events`.
pub(crate) fn call_llm_json(
    system: Option<&str>,
    user: &str,
    config: &crate::config::ZtoolsConfig,
) -> Option<serde_json::Value> {
    let model = resolve_weekend_model(&config.osaurus_url, &config.weekend_model);
    let content = crate::ztools::llm::chat(
        &crate::ztools::llm::ChatRequest {
            base_url: &config.osaurus_url,
            model: &model,
            system,
            user,
            json: true,
        },
        &config.chat_budget(config.llm_extended_timeout_secs),
    )
    .ok()?;
    Some(serde_json::json!({"choices": [{"message": {"content": content}}]}))
}

/// Condense a forecast to 1-2 sentences; fall back to a preview on failure.
#[must_use]
pub fn condense_weather(weather_str: &str, config: &crate::config::ZtoolsConfig) -> String {
    let prompt = render(PHASE_WEATHER_CONDENSE, &[("weather_str", weather_str)]);
    call_llm_text(&prompt, config)
        .unwrap_or_else(|| weather_str.chars().take(WEATHER_PREVIEW_LIMIT).collect())
}

/// Phase 1: pull clean pipe-separated lines out of the raw scraped corpus.
///
/// Batched with adaptive sizing: a streak of successes grows the batch, a
/// failure halves it and (at batch 1) falls back to passing the line through
/// raw rather than dropping it. The Python original persists batch sizes to a
/// signals file; this port keeps them in-memory per run, which is what the
/// shapes actually depend on.
#[must_use]
pub fn extract_sources(
    raw_text: &str,
    location: &str,
    config: &crate::config::ZtoolsConfig,
) -> String {
    let raw_lines: Vec<&str> = raw_text
        .lines()
        .filter(|l| {
            let t = l.trim_start();
            t.starts_with("- ") || t.starts_with("[THIS WEEKEND]")
        })
        .collect();
    if raw_lines.is_empty() {
        return raw_text.to_string();
    }

    let (marked, general): (Vec<&str>, Vec<&str>) = raw_lines
        .into_iter()
        .partition(|l| l.trim_start().starts_with("[THIS WEEKEND]"));
    let mut lines = marked;
    lines.extend(general.into_iter().take(24));

    let mut results = Vec::new();
    let mut batch_size = DEFAULT_BATCH_SIZE;
    let mut streak = 0;
    let mut failures = 0;
    let mut i = 0;
    while i < lines.len() {
        if failures >= EXTRACT_FAILURE_BREAKER {
            eprintln!(
                "\u{26a0} extract: {failures} consecutive model failures; passing the \
                 remaining {} line(s) through raw",
                lines.len() - i
            );
            results.extend(lines[i..].iter().map(ToString::to_string));
            break;
        }
        let end = cmp::min(i + batch_size, lines.len());
        let chunk = lines[i..end].join("\n");
        let prompt = render(
            PHASE_EXTRACT_EVENTS,
            &[("location", location), ("raw_text", &chunk)],
        );
        if let Some(res) = call_llm_text(&prompt, config) {
            results.push(res);
            streak += 1;
            failures = 0;
            i = end;
            if streak >= BATCH_GROWTH_STREAK_LIMIT && batch_size < MAX_BATCH_SIZE {
                batch_size += 1;
            }
        } else {
            streak = 0;
            failures += 1;
            batch_size = cmp::max(batch_size / 2, 1);
            if batch_size == 1 {
                // A single line that even the model rejects is passed through
                // rather than dropped: an empty extract is worse than a raw one.
                results.push(lines[i].to_string());
                i += 1;
                batch_size = DEFAULT_BATCH_SIZE;
            }
        }
    }

    if results.is_empty() {
        raw_text.to_string()
    } else {
        results.join("\n")
    }
}

/// Phase 2: draft candidate activities from the cleaned sources.
#[must_use]
pub fn draft_activities(
    weather_condensed: &str,
    cleaned_sources: &str,
    ctx: &PlanContext,
    config: &crate::config::ZtoolsConfig,
) -> Option<String> {
    let prompt = render(
        PHASE_DRAFT_TRANSIENT,
        &[
            ("age_range", &ctx.ages),
            ("location", &ctx.location),
            ("date_range", &ctx.date_range),
            ("year", &ctx.year.to_string()),
            ("weather_condensed", weather_condensed),
            ("cleaned_sources", cleaned_sources),
            ("carry", CARRY_FIELDS),
            ("exclusions", &ctx.exclusions),
        ],
    );
    call_llm_text(&prompt, config)
}

/// Phase 3: merge near-duplicates, keep the best, sort by appeal.
#[must_use]
pub fn refine_draft(draft_text: &str, config: &crate::config::ZtoolsConfig) -> String {
    let prompt = render(PHASE_REFINE, &[("draft_text", draft_text)]);
    call_llm_text(&prompt, config).unwrap_or_else(|| draft_text.to_string())
}

/// Phase 4: structure the refined draft into the transient-event JSON schema.
#[must_use]
pub fn structure_to_json(
    text: &str,
    weather_condensed: &str,
    year: i32,
    config: &crate::config::ZtoolsConfig,
) -> Option<Vec<super::WeekendEvent>> {
    let sys = render(
        PHASE_STRUCTURE_TRANSIENT_SYSTEM,
        &[
            ("year", &year.to_string()),
            ("weather_condensed", weather_condensed),
        ],
    );
    let usr = render(PHASE_STRUCTURE_USER, &[("draft_text", text)]);
    let resp = call_llm_json(Some(&sys), &usr, config)?;
    super::parse_llm_events(&resp)
}
